### PR TITLE
Bug fixes and optimisations

### PR DESCRIPTION
**Pull request by ksandom**
* Fix missing events.
* Fix unit test OOM:
  * TL;DR I had a huge number of config items being generated for all the possible combinations of criteria that can generate an event. Only a few of these are actually used at one time.
  * No longer generate config items for every possible event.
  * Add config templates that generate the config items for these scenarios for only the needed config items. These templates are chosen based on a regex matching against the config item name. They provide a description, and a default value, which can then be overridden as any other config item can.
  * I've also applied this fix to other areas where config items were generated in a for loop.
* Refactor load/save config code to make it easier to debug and refactor.
* Improve code quality.

### CODE REVIEW

**Comment by ksandom** [comment]
I forgot to also note that:

* Startup times are now noticeably faster.
* Memory usage is lower.
